### PR TITLE
Allow using arbitrary edge features in GNNs

### PR DESCRIPTION
**Pull request by mallamanis**

### CODE REVIEW

File: ptgnn/baseneuralmodel/abstractneuralmodel.py [hunk 1]
```diff
@@ -152,7 +152,7 @@ def build_neural_module(self) -> TNeuralModule:
 
     # region Saving/Loading
     def save(self, path: Path, model: TNeuralModule) -> None:
-        os.makedirs(os.path.dirname(str(path)), exist_ok=True)
+        os.makedirs(os.path.dirname(str(path.absolute())), exist_ok=True)
```
Comment (mmjb) [inline ptgnn/baseneuralmodel/abstractneuralmodel.py:155]: unrelated change?

File: ptgnn/implementations/typilus/train.py [hunk 2]
```diff
@@ -149,9 +150,12 @@ def run(arguments):
 
     initialize_metadata = True
     restore_path = arguments.get("--restore-path", None)
-    if restore_path or (arguments["--aml"] and model_path.exists()):
+    if restore_path:
         initialize_metadata = False
         model, nn = Graph2Class.restore_model(Path(restore_path))
+    elif arguments["--aml"] and model_path.exists():
```
Comment (mmjb) [inline ptgnn/implementations/typilus/train.py:156]: unrelated change?

File: ptgnn/neuralmodels/gnn/graphneuralnetwork.py [hunk 3]
```diff
@@ -88,12 +95,29 @@ def gnn(
...
+                adjacency_lists, edge_feature_embeddings
+            ):
+                mask = (
+                    torch.rand_like(edge_sources_idxs, dtype=torch.float32)
+                    > self.__edge_dropout_rate
+                )
+                dropped_adj_list.append(
+                    (edge_sources_idxs.masked_select(mask), edge_target_idxs.masked_select(mask))
+                )
+                dropped_edge_features.append(edge_features[mask])
```
Comment (mmjb) [inline ptgnn/neuralmodels/gnn/graphneuralnetwork.py:117]: The standard dropout implementation usually also does rescaling (see e.g. https://pytorch.org/docs/stable/generated/torch.nn.Dropout.html). Do that here as well?

File: ptgnn/neuralmodels/gnn/graphneuralnetwork.py [hunk 4]
```diff
@@ -257,25 +323,39 @@ def __iterate_edge_types(
...
+            return None
+
+        if self.__edge_embedding_model is None:
+            tensorized_edge_features = None
+        else:
+            tensorized_edge_features = []
+            for edge_type in self.__edge_idx_to_type:
+                edge_features = datapoint.edge_features.get(edge_type)
+                if edge_features is None:
+                    tensorized_edge_features.append([])
```
Comment (mmjb) [inline ptgnn/neuralmodels/gnn/graphneuralnetwork.py:340]: I'm a little bit confused by this one - the list is never consumed, right? Would it be better to use `None` here to make clear that this data is missing? (`[]` for "no data" can have weird effects when it's accessed using `for` or `zip` when it silently truncates instead of failing)

File: ptgnn/neuralmodels/gnn/graphneuralnetwork.py [hunk 5]
```diff
@@ -306,16 +394,35 @@ def extend_minibatch_with(
         graph_idx = len(partial_minibatch["num_nodes_per_graph"])
 
         adj_list = partial_minibatch["adjacency_lists"]
+        tensorized_edge_feature_data = partial_minibatch["edge_feature_data"]
         nodes_in_mb_so_far = partial_minibatch["num_nodes_in_mb"]
-        for sample_adj_list_for_edge_type, mb_adj_lists_for_edge_type in zip(
-            tensorized_datapoint.adjacency_lists, adj_list
+
+        all_edge_features = tensorized_datapoint.edge_features
```
Comment (mmjb) [inline ptgnn/neuralmodels/gnn/graphneuralnetwork.py]: The naming here is super-confusing: `all_edge_features` is all edge features of the considered datapoint, but `tensorized_edge_feature_data` is all edge features in the minibatch. Maybe rename for clarity?

**Review by mmjb — COMMENTED** [review]
In principle this seems fine (see my minor comments). One thing I was wondering about was if it would make sense to support the case of having edge features only for some edges explicitly, but it's probably easier to ignore this for now (or just use constant values for those edges)

File: ptgnn/baseneuralmodel/abstractneuralmodel.py (hunk 1 above)
Comment (mallamanis) [reply ptgnn/baseneuralmodel/abstractneuralmodel.py:155]: yes...

File: ptgnn/implementations/typilus/train.py (hunk 2 above)
Comment (mallamanis) [reply ptgnn/implementations/typilus/train.py:156]: yes, again (allows pre-empted workloads to resume)

File: ptgnn/neuralmodels/gnn/graphneuralnetwork.py (hunk 5 above)
Comment (mallamanis) [reply ptgnn/neuralmodels/gnn/graphneuralnetwork.py]: Correct. The next commit will correct the variable name.

File: ptgnn/neuralmodels/gnn/graphneuralnetwork.py (hunk 4 above)
Comment (mallamanis) [reply ptgnn/neuralmodels/gnn/graphneuralnetwork.py:340]: Thanks, this was a bit confusing indeed. 

I've renamed `edge_features` to `edge_feature_for_edge_type`. `edge_features` is `None` when there are no edges of the given type in the datapoint, not when there are no edge features.

`tensorized_edge_features` should be None when features aren't used (= no `edge_embedding_model`)

Comment (mmjb) [reply ptgnn/neuralmodels/gnn/graphneuralnetwork.py:340]: Ah! I indeed didn't get this, and the comment now helps as well.

**Review by mmjb — APPROVED** [review]
